features/graph/api/src/main/java/org/opennms/netmgt/graph/api/generic/AllowedValuesInPropertiesMap.java

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * This file is part of OpenNMS(R).
+ *
+ * Copyright (C) 2019-2019 The OpenNMS Group, Inc.
+ * OpenNMS(R) is Copyright (C) 1999-2019 The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is a registered trademark of The OpenNMS Group, Inc.
+ *
+ * OpenNMS(R) is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * OpenNMS(R) is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with OpenNMS(R).  If not, see:
+ *      http://www.gnu.org/licenses/
+ *
+ * For more information contact:
+ *     OpenNMS(R) Licensing <license@opennms.org>
+ *     http://www.opennms.org/
+ *     http://www.opennms.com/
+ *******************************************************************************/
+
+package org.opennms.netmgt.graph.api.generic;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import org.opennms.netmgt.graph.api.info.NodeInfo;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+/** 
+ * This class holds all allowed types that can be used inside the properties of a GenericElement. They need to be immutable and persistable.
+ * See also org.opennms.netmgt.graph.persistence.converter.ConverterService
+ * TODO: Patrick: discuss with mvr if and how we want to synchronize this list with the converters in converter service. 
```

Comment:
I introduced this class in order to ensure immutability in values of the properties map. Since java doesn't offer a built in way to ensure immutability I see two ways to solve this:
1.  explicitly allow the allow types (as implemented here). This is a fast straight forward approach but needs to be extended each time if we need to add different type. I think this should be ok since we anyway need to extend the converters in the persistence layer for each new type anyway. It would be nice to synchronize those 2 lists somehow.
2. make a generic solution to detect immutability (allowed type could then be cached for speed). Here is an example how this could be achieved: https://github.com/MutabilityDetector/MutabilityDetector4FindBugs

@mvrueden, lets have a discussion which way we want to go.

Reply:
Given that the Map's value type is `Object` achieving deep immutability is likely not worth the effort.

Could we just simply state via javadoc that all elements of this collection must be immutable and persistable and then drop the explicit checking?

Let me know if this functionality is must have.

Reply:
In order to persist a value we use this converter service: https://github.com/OpenNMS/opennms/blob/features/graph-service/features/graph/service/src/main/java/org/opennms/netmgt/graph/persistence/converter/ConverterService.java
which limits the types of allowed values to a handful. The question is if it doesn't makes sense to fail already at construction time instead at persisting time?

Reply:
Mainly each Graph must be serialisable (to JSON) and may be persisted to the database.
We should only enforce the property type if actually persisted to the database.
Besides that, as already mentioned let's not worry too much about deep cloning, as we don't really have that problem.